polynote-kernel/src/main/scala/polynote/kernel/remote/transport.scala

```diff
@@ -198,10 +200,23 @@ class SocketTransport(
   forceServerAddress: Option[String] = None
 ) extends Transport[InetSocketAddress] {
 
-  private def openServerChannel: RIO[Blocking, ServerSocketChannel] = effectBlocking {
-    ServerSocketChannel.open().bind(
-      new InetSocketAddress(
-        forceServerAddress.getOrElse(java.net.InetAddress.getLocalHost.getHostAddress), 0))
+  private[this] def socketIsFree(host: String, port: Int): Boolean =
```

Comment:
So the idea is that if we can't connect to a port then it is available?... maybe it would be simpler to just repeatedly try to bind, inside of `openServerChannel`? Something like:

```scala
val channel = ServerSocketChannel.open()
config.kernel.portRange.fold(effectBlocking(channel.bind(address, 0))) {
  range => effectBlocking(channel.bind(address, scala.util.Random.nextInt(range.size) + range.start))
    .retry(Schedule.recurs(20))
}
```

(or some other way to sweep over the ports)

Reply:
Yes that's the idea. I like the way you wrote it better. I'm changing it.

Reply:
I am having trouble with types, the fold expect the types to be identical but retry add the trait `Clock` to `Blocking`.

Reply:
That should be OK, just add `Clock` to the return type's environment. Upstream probably has `Clock`.